orttraining/orttraining/training_api/onnxruntime_training_c_api.cc

```diff
@@ -9,6 +9,20 @@
 #include "orttraining/training_api/include/training_session.h"
 #include "core/session/abi_session_options_impl.h"
 
+namespace {
+
+std::vector<std::shared_ptr<onnxruntime::IExecutionProvider>> CreateProviders(
+    const std::vector<std::shared_ptr<onnxruntime::IExecutionProviderFactory>>& provider_factories) {
+  std::vector<std::shared_ptr<onnxruntime::IExecutionProvider>> execution_providers;
+  for (const auto& factory : provider_factories) {
+    execution_providers.emplace_back(std::move(factory->CreateProvider()));
```

Comment:
Since we are creating the provider here it means the same instance of the provider will be shared among training, eval and optimizer session right? 

For inference scenarios we don't share ep instance among inference sessions but in ortmodule we do... Just wondering do we know any implications of sharing the provider instance?

Reply:
I am not sure if there are any guidelines around sharing the provider among different inference sessions. If there is, please let me know.

I think from a programming viewpoint, we should think of `TrainingSession` as an equivalent to the `InferenceSession` just in the training world. Extending that further, all components inside the `TrainingSession` should share the same instance of the provider and by extension the allocator (per provider). But if this is not the expected usage for the provider, we can change it.

@pengwa, @ashbhandare please provide any insight that may be relevant.